internal/langserver/handlers/command/module_list.go

```diff
@@ -0,0 +1,145 @@
+	Version          string            `json:"version"`
+	Type             string            `json:"type"`
+	DocLink          string            `json:"docklink"`
+	DependentModules []terraformModule `json:"depmodules"`
+}
+
+func ModuleListHandler(ctx context.Context, args cmd.CommandArgs) (interface{}, error) {
+	modUri, ok := args.GetString("uri")
+	if !ok || modUri == "" {
+		return nil, fmt.Errorf("%w: expected module uri argument to be set", code.InvalidParams.Err())
+	}
+
+	if !uri.IsURIValid(modUri) {
+		return nil, fmt.Errorf("URI %q is not valid", modUri)
+	}
+
+	modPath, err := uri.PathFromURI(modUri)
+	if err != nil {
+		return nil, err
+	}
+
+	mm, err := lsctx.ModuleFinder(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	found, _ := mm.ModuleByPath(modPath)
+	if found.ModManifest == nil {
+		return nil, nil
+	}
+
+	list := parseModuleRecords(found.ModManifest.Records)
+
+	return moduleListResponse{
+		Modules:       list,
+		FormatVersion: moduleListVersion,
+	}, nil
+}
+
+func parseModuleRecords(records []datadir.ModuleRecord) []terraformModule {
+	modules := make(map[string]terraformModule)
+	for _, manifest := range records {
+		if manifest.Dir == "." {
+			// this is the current directory, which is technically a module
+			// skipping as it's not relevant in the activity bar (yet?)
+			continue
+		}
+
+		moduleName := manifest.Key
+		subModuleName := ""
+
+		// determine if this module is nested in another module
+		// in the currecnt workspace by finding a period in the moduleName
+		// is it better to look at SourceAddr and compare?
+		if strings.Contains(manifest.Key, ".") {
+			v := strings.Split(manifest.Key, ".")
+			moduleName = v[0]
+			subModuleName = v[1]
+		}
+
+		// github modules are put in `files` directory, but Manifest Record does not
+		// record the `name` in Key it puts it in SourceAddr
+		if strings.HasPrefix(manifest.Key, "files") {
+			moduleName = manifest.SourceAddr
+		}
+
+		// build what we know
+		moduleInfo := terraformModule{
+			Name:    moduleName,
+			Path:    strings.TrimSuffix(manifest.SourceAddr, "/"),
+			Version: manifest.VersionStr,
+			Type:    getModuleType(manifest),
+		}
+
+		if moduleInfo.Type == "tfregistry" {
+			moduleInfo.DocLink = getModuleDocumentationLink(manifest)
+		}
+
+		m, present := modules[moduleName]
+		if present {
+			// this module is located inside another so append
+			moduleInfo.Name = subModuleName
+			m.DependentModules = append(m.DependentModules, moduleInfo)
+			modules[moduleName] = m
+		} else {
+			// this is the first we've seen module
+			modules[moduleName] = moduleInfo
+		}
+	}
+
+	// don't need the map anymore, return a list of modules found
+	list := make([]terraformModule, 0)
+	for _, mo := range modules {
+		list = append(list, mo)
+	}
+
+	sort.SliceStable(list, func(i, j int) bool {
+		return list[i].Name < list[j].Name
+	})
+
+	return list
+}
+
+func getModuleType(manifest datadir.ModuleRecord) string {
```

Comment:
It may be worth decoupling this into `internal/terraform/datadir/module_types.go` and relatedly turn the known types into constants.

These can be declared on `*ModuleRecord`.

I think that longer-term we might need to extract it out of the repo entirely, similar to https://github.com/hashicorp/terraform-registry-address but for now keeping it in a separate package should suffice. I can consider the extraction as part of https://github.com/hashicorp/terraform-ls/issues/599 and/or https://github.com/hashicorp/terraform-ls/issues/598

Reply:
Moved to new file `internal/terraform/datadir/module_types.go` in this iteration